frame/support/src/storage/mod.rs

```diff
@@ -811,16 +811,130 @@ mod private {
 
 	impl<T: Encode> Sealed for Vec<T> {}
 	impl<Hash: Encode> Sealed for Digest<Hash> {}
+	impl<T: ValueType, S: Get<usize>> Sealed for BoundedVec<T, S> {}
 }
 
 impl<T: Encode> StorageAppend<T> for Vec<T> {}
 impl<T: Encode> StorageDecodeLength for Vec<T> {}
 
-/// We abuse the fact that SCALE does not put any marker into the encoding, i.e.
-/// we only encode the internal vec and we can append to this vec. We have a test that ensures
-/// that if the `Digest` format ever changes, we need to remove this here.
+/// We abuse the fact that SCALE does not put any marker into the encoding, i.e. we only encode the
+/// internal vec and we can append to this vec. We have a test that ensures that if the `Digest`
+/// format ever changes, we need to remove this here.
 impl<Hash: Encode> StorageAppend<DigestItem<Hash>> for Digest<Hash> {}
 
+use crate::traits::Get;
+use sp_std::convert::TryFrom;
+
+pub trait ValueType: FullCodec + Default {}
+impl<T: FullCodec + Default> ValueType for T {}
+
+#[derive(Encode, Decode, Default)]
+pub struct BoundedVec<T: ValueType, S: Get<usize>>(Vec<T>, sp_std::marker::PhantomData<S>);
+
+impl<T: ValueType, S: Get<usize>> BoundedVec<T, S> {
+	fn unchecked_from(t: Vec<T>) -> Self {
+		Self(t, Default::default())
+	}
+	// TODO: add a subset of the Vec's API here that generally doesn't expand it in any way. Adding
+	// any API that can insert/append/enqueue is tricky and needs care.
+}
+
+impl<T: ValueType, S: Get<usize>> TryFrom<Vec<T>> for BoundedVec<T, S> {
+	type Error = ();
+	fn try_from(t: Vec<T>) -> Result<Self, Self::Error> {
+		if t.len() <= S::get() {
+			Ok(Self::unchecked_from(t))
+		} else {
+			Err(())
+		}
+	}
+}
+
+// It is okay to give a non-mutable reference of the inner vec to anyone.
+impl<T: ValueType, S: Get<usize>> AsRef<Vec<T>> for BoundedVec<T, S> {
+	fn as_ref(&self) -> &Vec<T> {
+		&self.0
+	}
+}
+
+impl<T: ValueType, S: Get<usize>> codec::DecodeLength for BoundedVec<T, S> {
+	fn len(mut self_encoded: &[u8]) -> Result<usize, codec::Error> {
+		use sp_std::convert::TryFrom;
+		// `BoundedVec<T, _>` stored just a `Vec<T>`, thus the length is at the beginning in
+		// `Compact` form, and same implementation as `Vec<T>` can be used.
+		<Vec<T> as codec::DecodeLength>::len(self_encoded)
+	}
+}
+
+impl<T: ValueType, S: Get<usize>> StorageDecodeLength for BoundedVec<T, S> {}
+
+pub trait TryAppend<T: ValueType, S: Get<usize>> {
+	fn try_append<LikeT: EncodeLike<T>>(item: LikeT) -> Result<(), ()>;
+}
+
+impl<T: ValueType, S: Get<usize>, V: generator::StorageValue<BoundedVec<T, S>>> TryAppend<T, S>
+	for V
+{
+	fn try_append<LikeT: EncodeLike<T>>(item: LikeT) -> Result<(), ()> {
```

Comment:
yeah, though eventually we'll probably want this to become a `DispatchError` to avoid a lot of `.map_err(|()| Error::<T>::ItemOverflow)`s.

Reply:
Not sure which variant it should be? `Other` or a new (general)`Overflow` would be okay, but the benefit of `.map_err(|()| Error::<T>::ItemOverflow)` is actually that we know from which pallet it is coming from.